src/test_diag.jl

```diff
@@ -0,0 +1,215 @@
+using ForwardDiff
+using Random
+using Distributions
+include("DiagonalHessianApproximation.jl")
+
+Random.seed!(123)
+d = Normal(0.0,1.0)
+dnoise = Normal(0.0,0.1)
+
+# Points
+x0 = rand(d,3)
+x1 = x0 + rand(dnoise,3)
+
+# Test functions
+f(x) = x[1]^2 + x[2]^2 * x[3]^2
+g(x) = exp(x[1]) + x[2] + cos(x[3])
+h(x) = x[1]^2 * x[2] * x[3]^3
+
+############################
+# Test DiagonalQN
+############################
+
+# Exact initialization
+Bf = DiagonalQN(diag(ForwardDiff.hessian(f,x0)))
+Bg = DiagonalQN(diag(ForwardDiff.hessian(g,x0)))
+Bh = DiagonalQN(diag(ForwardDiff.hessian(h,x0)))
+
+# Update the approximations
+push!(Bf,
+      x1-x0,
+      ForwardDiff.gradient(f,x1)-ForwardDiff.gradient(f,x0))
+push!(Bg,
+      x1-x0,
+      ForwardDiff.gradient(g,x1)-ForwardDiff.gradient(g,x0))
+push!(Bh,
+      x1-x0,
+      ForwardDiff.gradient(h,x1)-ForwardDiff.gradient(h,x0))
+
+
+println("\nTEST DiagonalQN\n")
+println("True f :")
+println(diag(ForwardDiff.hessian(f,x1)))
+println("Approx f :")
+println(diag(Matrix(Bf)))
+println("True g :")
+println(diag(ForwardDiff.hessian(g,x1)))
+println("Approx g :")
+println(diag(Matrix(Bg)))
+println("True h :")
+println(diag(ForwardDiff.hessian(h,x1)))
+println("Approx h :")
+println(diag(Matrix(Bh)))
+
+############################
+# Test SpectralGradient
+############################
+
+# Exact initialization
+Bf = SpectralGradient(diag(ForwardDiff.hessian(f,x0)))
+Bg = SpectralGradient(diag(ForwardDiff.hessian(g,x0)))
+Bh = SpectralGradient(diag(ForwardDiff.hessian(h,x0)))
+
+# Update the approximations
+push!(Bf,
+      x1-x0,
+      ForwardDiff.gradient(f,x1)-ForwardDiff.gradient(f,x0))
+push!(Bg,
+      x1-x0,
+      ForwardDiff.gradient(g,x1)-ForwardDiff.gradient(g,x0))
+push!(Bh,
+      x1-x0,
+      ForwardDiff.gradient(h,x1)-ForwardDiff.gradient(h,x0))
+
+println("\nTEST SpectralGradient\n")
+println("True f :")
+println(diag(ForwardDiff.hessian(f,x1)))
+println("Approx f :")
+println(diag(Matrix(Bf)))
+println("True g :")
+println(diag(ForwardDiff.hessian(g,x1)))
+println("Approx g :")
+println(diag(Matrix(Bg)))
+println("True h :")
+println(diag(ForwardDiff.hessian(h,x1)))
+println("Approx h :")
+println(diag(Matrix(Bh)))
```

Comment:
As we discussed:

- don't print in unit tests; everything should be testable automatically;
- use `@testset` to group related tests together.